File: my-reader/src-tauri/src/sync/annotation.rs
use sea_orm::DatabaseConnection;
use uuid::Uuid;

use crate::entities::app::annotations;
use crate::error::AppError;
use crate::repositories::annotation_repo::SqliteAnnotationRepository;

use super::automerge_document::{
    annotation_projections, create_annotation, delete_annotation, update_annotation,
    AnnotationValue,
};
use super::automerge_projection::LibrarySidecarAutomergeProjection;
use super::automerge_store::commit_library_sidecar_automerge_mutation;
use super::replica_identity::ensure_replica_identity;

#[allow(clippy::too_many_arguments)]
pub async fn add_local_annotation(
    db: &DatabaseConnection,
    library_uuid: &str,
    book_id: i64,
    format: &str,
    locator_json: &str,
    color: &str,
    note: Option<&str>,
    now_ms: u64,
) -> Result<annotations::Model, AppError> {
    let identity = ensure_replica_identity(db, library_uuid).await?;
    let id = Uuid::new_v4().as_simple().to_string();
    let projection = LibrarySidecarAutomergeProjection;
    commit_library_sidecar_automerge_mutation(
        db,
        &identity,
        now_ms,
        |document| {
            create_annotation(
                document,
                &AnnotationValue {
                    id: id.clone(),
                    book_id,
                    format: format.to_owned(),
                    kind: "highlight".to_owned(),
                    locator_json: locator_json.to_owned(),
                    created_at: now_ms as i64,
                    color: color.to_owned(),
                    note: note.map(ToOwned::to_owned),
                    updated_at: now_ms as i64,
                    deleted: false,
                    deleted_at: None,
                },
            )?;
            Ok(())
        },
        Some(&projection),
    )
    .await?;
    SqliteAnnotationRepository::find_by_id(db, &id)
        .await?
        .ok_or_else(|| AppError::Database("Annotation creation returned no row".into()))
}

#[allow(clippy::too_many_arguments)]
pub async fn update_local_annotation(
    db: &DatabaseConnection,
    library_uuid: &str,
    id: &str,
    book_id: i64,
    format: &str,
    color: &str,
    note: Option<&str>,
    now_ms: u64,
) -> Result<Option<annotations::Model>, AppError> {
    let identity = ensure_replica_identity(db, library_uuid).await?;
    let projection = LibrarySidecarAutomergeProjection;
    let mut exists = true;
    commit_library_sidecar_automerge_mutation(
        db,
        &identity,
        now_ms,
        |document| {
            let current = annotation_projections(document)?
                .into_iter()
                .find(|annotation| {
                    annotation.id == id
                        && annotation.book_id == book_id
                        && annotation.format == format
                        && !annotation.deleted
                });
            if current.is_none() {
                exists = false;
                return Ok(());
            }
            update_annotation(document, id, color, note, now_ms as i64)?;
            Ok(())
        },
        Some(&projection),
    )
    .await?;
    if exists {
        SqliteAnnotationRepository::find_by_id(db, id).await
    } else {
        Ok(None)
    }
}

pub async fn delete_local_annotation(
    db: &DatabaseConnection,
    library_uuid: &str,
    id: &str,
    book_id: i64,
    format: &str,
    now_ms: u64,
) -> Result<bool, AppError> {
    let identity = ensure_replica_identity(db, library_uuid).await?;
    let projection = LibrarySidecarAutomergeProjection;
    let mut exists = true;
    commit_library_sidecar_automerge_mutation(
        db,
        &identity,
        now_ms,
        |document| {
            let current = annotation_projections(document)?
                .into_iter()
                .find(|annotation| {
                    annotation.id == id
                        && annotation.book_id == book_id
                        && annotation.format == format
                        && !annotation.deleted
                });
            if current.is_none() {
                exists = false;
                return Ok(());
            }
            delete_annotation(document, id, now_ms as i64)?;
            Ok(())
        },
        Some(&projection),
    )
    .await?;
    Ok(exists)
}

File: my-reader-mobile/src/domain/sync/library-sidecar/annotation.ts
import type { Library } from "@my-reader/tools/types/library"

import {
  readLibrarySidecarAnnotation,
  withLibrarySidecarSyncTransaction,
  type LibrarySidecarAnnotationRow,
} from "@/src/repos/library-sidecar-sync"
import {
  librarySidecarAnnotationProjections,
  type LibrarySidecarAnnotationValue,
} from "./automerge-document"
import { commitLibrarySidecarAutomergeMutation } from "./automerge-store"
import { ensureLibrarySidecarIdentity } from "./identity"

async function projectedAnnotation(
  library: Library,
  id: string,
): Promise<LibrarySidecarAnnotationRow | null> {
  return withLibrarySidecarSyncTransaction(library, (tx) =>
    readLibrarySidecarAnnotation(tx, id),
  )
}

export async function createLocalAnnotation(
  library: Library,
  value: Omit<
    LibrarySidecarAnnotationValue,
    "createdAt" | "updatedAt" | "deleted" | "deletedAt"
  >,
  nowMs = Date.now(),
): Promise<LibrarySidecarAnnotationRow> {
  const identity = await ensureLibrarySidecarIdentity(library)
  await commitLibrarySidecarAutomergeMutation(library, identity, nowMs, () => ({
    type: "createAnnotation",
    value: {
      ...value,
      createdAt: nowMs,
      updatedAt: nowMs,
      deleted: false,
      deletedAt: null,
    },
  }))
  const row = await projectedAnnotation(library, value.id)
  if (!row) throw new Error("Annotation creation returned no row")
  return row
}

export async function updateLocalAnnotation(
  library: Library,
  id: string,
  color: string,
  note: string | null,
  nowMs = Date.now(),
): Promise<LibrarySidecarAnnotationRow | null> {
  const identity = await ensureLibrarySidecarIdentity(library)
  let exists = true
  await commitLibrarySidecarAutomergeMutation(
    library,
    identity,
    nowMs,
    (document) => {
      const current = librarySidecarAnnotationProjections(document).find(
        (annotation) => annotation.id === id && !annotation.deleted,
      )
      if (!current) {
        exists = false
        return null
      }
      return { type: "updateAnnotation", id, color, note, updatedAt: nowMs }
    },
  )
  return exists ? projectedAnnotation(library, id) : null
}

export async function deleteLocalAnnotation(
  library: Library,
  id: string,
  nowMs = Date.now(),
): Promise<boolean> {
  const identity = await ensureLibrarySidecarIdentity(library)
  let exists = true
  await commitLibrarySidecarAutomergeMutation(
    library,
    identity,
    nowMs,
    (document) => {
      const current = librarySidecarAnnotationProjections(document).find(
        (annotation) => annotation.id === id && !annotation.deleted,
      )
      if (!current) {
        exists = false
        return null
      }
      return { type: "deleteAnnotation", id, deletedAt: nowMs }
    },
  )
  return exists
}
